vowpalwabbit/cache.h

```diff
@@ -19,17 +19,16 @@ struct cache_temp_buffer;
 char* run_len_decode(char* p, size_t& i);
 char* run_len_encode(char* p, size_t i);
 
-int read_cached_features(VW::workspace* all, io_buf& buf, v_array<example*>& examples);
 void cache_tag(io_buf& cache, const v_array<char>& tag);
-void cache_features(io_buf& cache, example* ae, uint64_t mask);
 void output_byte(io_buf& cache, unsigned char s);
-void output_features(io_buf& cache, unsigned char index, features& fs, uint64_t mask);
+void cache_index_and_features(io_buf& cache, unsigned char index, features& fs, uint64_t mask);
+std::pair<size_t, bool> read_cached_index_and_features(io_buf& input, unsigned char& index, features& ours);
 
 namespace VW
 {
 uint32_t convert(size_t number);
 // What is written by write_example_to_cache can be read by read_example_from_cache
 void write_example_to_cache(io_buf& output, example* ae, label_parser& lbl_parser, uint64_t parse_mask,
     VW::details::cache_temp_buffer& temp_buffer);
-int read_example_from_cache(io_buf& input, example* ae, label_parser& lbl_parser, bool sorted_cache);
+int read_example_from_cache(VW::workspace* all, io_buf& buf, v_array<example*>& examples);
```

Comment:
why do we need to pass in `all` here? In general it's best to avoid passing the all object when it's not needed, and passing in more specific args to functions is desirable, so I am wondering why we are going from `io_buf` to `all`?

Reply:
This function definition is required because it's passed in as a function pointer to the parser. This is replacing the old `read_cached_features`